parlai/agents/transformer/polyencoder.py

```diff
@@ -0,0 +1,267 @@
+from .modules import BasicAttention, MultiHeadAttention
+import torch
+
+class PolyencoderAgent(TorchRankerAgent):
+    """ Equivalent of bert_ranker/polyencoder and biencoder_multiple_output
+        but does not rely on an external library (hugging face).
+    """
+
+    @classmethod
+    def add_cmdline_args(cls, argparser):
+        """Add command-line arguments specifically for this agent."""
+        TransformerRankerAgent.add_cmdline_args(argparser)
+        agent = argparser.add_argument_group('Polyencoder Arguments')
+        agent.add_argument('--polyencoder-type', type=str, default='codes',
+                           choices=['codes', 'n_first'],
+                           help='Type of polyencoder, either we compute'
+                                'vectors using codes + attention, or we '
+                                'simply take the first N vectors.')
+        agent.add_argument('--poly-n-codes', type=int, default=64,
+                           help='number of vectors used to represent the context'
+                                'in the case of n_first, those are the number'
+                                'of vectors that are considered.')
+        agent.add_argument('--poly-attention-type', type=str, default='basic',
+                           choices=['basic', 'basic_sqrt', 'multihead'],
+                           help='Type of the top aggregation layer of the poly-'
+                                'encoder (where the candidate representation is'
+                                'the key)')
+        agent.add_argument('--poly-attention-num-heads', type=int, default=4,
+                           help='In case poly-attention-type is multihead, '
+                                'specify the number of heads')
+
+        # Those arguments are here in case where polyencoder type is 'code'
+        agent.add_argument('--codes-attention-type', type=str, default='basic',
+                           choices=['basic', 'basic_sqrt', 'multihead'],
+                           help='Type ')
+        agent.add_argument('--codes-attention-num-heads', type=int, default=4,
+                           help='In case codes-attention-type is multihead, '
+                                'specify the number of heads')
+        return agent
+
+    def __init__(self, opt, shared=None):
+        super().__init__(opt, shared)
+        self.rank_loss = torch.nn.CrossEntropyLoss(reduce=True, size_average=True)
+        if self.use_cuda:
+            self.rank_loss.cuda()
+        self.data_parallel = opt.get('data_parallel') and self.use_cuda
+        if self.data_parallel:
+            from parlai.core.distributed_utils import is_distributed
+            if is_distributed():
+                raise ValueError(
+                    'Cannot combine --data-parallel and distributed mode'
+                )
+            self.model = torch.nn.DataParallel(self.model)
+        self.START_IDX = self.END_IDX
```

Comment:
i'd be careful about this call. not sure it will mess anything up, but you might want to fix the dictionary

Reply:
The same effect can be obtained with `--dict-endtoken __start__`